### PR TITLE
Upgrade to Rails 5.0

### LINKED ISSUE #1235: Upgrade from 4.2 to 5.0
### Resources:

- [Rails 5.0 release notes](https://guides.rubyonrails.org/5_0_release_notes.html)
- [Upgrading from Rails 4.2 to Rails 5.0](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#upgrading-from-rails-4-2-to-rails-5-0)

### [Prepare for Rails Upgrade](https://www.fastruby.io/blog/rails/upgrade/prepare-for-rails-upgrade.html)

- [x] Cleanup unused (and underused) gems (see #1511)
- [x] Code Coverage >= 80%, but consider:
  - [ ] adding specs for controllers at 0%
  - [ ] adding specs for controllers < 50%
  - [ ] adding specs for controllers < 80%
- [x] Staging and Production
- [x] Upgrade to latest patch release
- [x] Incompatabilities —[RailsBump](https://railsbump.org/lockfiles/c56eb7d3496ad0baf0a18d10fda215232a1e03a2) doesn't report much
  - [x] `devise_security_extensions` -> [`devise-security`](https://github.com/devise-security/devise-security)
- [x] Dual Boot (#1322)

### [Upgrade from Rails 4.2 to 5.0](https://www.fastruby.io/blog/rails/upgrades/upgrade-rails-from-4-2-to-5-0.html)

- [x] ~~Upgrade to Ruby 2.2~~ we're already on 2.2
- [x] Upgrade gems
- [x] `rails app:update`
- [x] Application Code
  - [x] Models
  - [x] Controllers
- [x] Testing
- [ ] Enable Rails 5.0 on Staging

### LINKED ISSUE #1322: Enable dual boot for having multiple Rails versions
The idea is to use a Gem that allows to have multiple versions of rails in the same project/branch.  You will select the version that you want to use at deployment time and it can be part of the CI.
This way, we will be able to work on new features and on migrating, both things in the same branch.  This helps avoiding the difficult situation of working on the migration on a branch that gets too behind the production branch, and then it gets super complicated to merge it.

Here's an article explaining how to enable the dual boot:
https://www.fastruby.io/blog/upgrade-rails/dual-boot/dual-boot-with-rails-6-0-beta.html

Here's the conversation about it:
https://manas.slack.com/archives/C034BR64J/p1630682303031000?thread_ts=1630571164.013600&cid=C034BR64J

### PR DESCRIPTION
**Pull request by ysbaddaden**
- [x] Dual boot current (Rails 4.0) or next (Rails 5.0) (closes #1322)
- [x] Upgrade gems:
  - [x] Lock current gems to versions that support Ruby 2.2
  - [x] Conservative upgrade of gems
  - [x] Upgrade to Devise 4.x
  - [x] Upgrade to Rails 5.0
  - [x] Upgrade to Sidekiq 4.2, and reenable Sidekiq::Web (in development)
  - [ ] ~~Upgrade to Sidekiq 5.0~~ for autoreload to work properly (postponed)
  - [x] Reenable disabled gems (e.g. `pry-byebug`)
  - [x] Rails 5 defaults to newer versions of these gems:
    - [x] Upgrade `jbuilder` to 2.5 (compatible with rails 4.2)
    - [ ] ~~Upgrade `turbolinks` to 5.0~~ postponed (e.g. #1236)
  - [ ] ~~Migrate from `therubyracer` to another `execjs` supported JS engine (e.g. node.js)~~ postponed (`mini_racer` caused integration specs to start failing)
- [x] Fix Rails 5.0 deprecations:
  - [x] `.relation(:reload)` is deprecated (use `relation.reload`)
  - [x] `alias_method_chain` is deprecated (use `prepend`)
  - [x] strong parameters issues (security)
  - [x] use `.distinct` instead of `.uniq`
  - [x] use `render json:` instead of the custom `render_json` helper
  - [x] migrate controller specs to new request DSL —maybe in a dedicated subsequent pull request? I do have a patch to use the DSL with Rails 4.2, but fixing all specs will take time
- [x] Fix broken unit tests
- [x] Fix broken integration tests
- [x] Dual boot: run CI tests for next on GHA
- [x] **BUG**: the 1st rspec integration test is always failing (Net::ReadTimeout).

I didn't enable the new framework defaults. We'll see that in subsequent developments. This is mostly because they would be breaking changes over Rails 4, and we wouldn't be able to run a Rails 5 version in production while being allowed to rollback to Rails 4.2 if needed.

Some dependencies seem to trigger some deprecation warnings (e.g. `react-rails` and `poirot-rails`), we'll probably have to upgrade —or fix them.

closes #1235

### CODE REVIEW

**Comment by ysbaddaden** [comment]
Dang, for some reasons the timed related specs in `spec/models/sample_transfer_spec.rb` started failing today, while they did pass on Friday :sob:

**Comment by ysbaddaden** [comment]
Sigh, a classic "milliseconds aren't retained", something changed in AR5 to reflect MySQL behavior (unless the column has some sub-second precision?):

```ruby
st = SampleTransfer.new
st.confirmed_at = Time.now
st.confirmed_at.nsec
# AR4 => 525905892
# AR5 => 0
```

File: app/models/encounter.rb [hunk 1]
```diff
@@ -27,6 +27,27 @@ def self.entity_scope
 
   attr_accessor :new_samples # Array({entity_id: String}) of new generated samples from UI.
 
+  # Finds an Encounter by its UUID or numerical ID.
+  def self.find_by_identifier(uuid_or_id)
+    if encounter = where(uuid: uuid_or_id).take
```
Comment (straight-shoota) [inline app/models/encounter.rb:32]: We should be able to differentiate `uuid_or_id` based on the UUID format. That should be sufficient to issue only a single DB request either way.

File: app/models/concerns/resource.rb [hunk 2]
```diff
@@ -71,7 +71,7 @@ def resource_name
       resource_name_prefix
     end
 
-    def filter(conditions)
+    def filter_with(conditions)
```
Comment (straight-shoota) [inline app/models/concerns/resource.rb:74]: What's the point of a single-use method which is just an alias for `where`?

Comment (ysbaddaden) [reply app/models/concerns/resource.rb]: :shrug:

**Comment by ysbaddaden** [comment]
Damn, there is a bug on CI with the Rails 5 gems, where the first rspec integration seems to always fail because of a `Net::TimeoutError`, then the following tests are passing... This is only affecting the rails5 runs.

**Pull request merged**